protokube/pkg/protokube/ali_volume.go

```diff
@@ -0,0 +1,322 @@
+	if ecsEndpoint == "" {
+		// TODO: shall we raise error here?
+		ecsEndpoint = ecs.ECSDefaultEndpoint
+	}
+
+	client := ecs.NewClientWithEndpoint(ecsEndpoint, accessKeyId, accessKeySecret)
+	a := &ALIVolumes{
+		client: client,
+	}
+
+	err := a.discoverTags()
+	if err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+// ClusterID implements Volumes ClusterID
+func (a *ALIVolumes) ClusterID() string {
+	return a.clusterTag
+}
+
+// InstanceID implements Volumes InstanceID
+func (a *ALIVolumes) InstanceID() string {
+	return a.instanceId
+}
+
+// InternalIP implements Volumes InternalIP
+func (a *ALIVolumes) InternalIP() net.IP {
+	return a.internalIP
+}
+
+func (a *ALIVolumes) discoverTags() error {
+	metadataClient := metadata.NewMetaData(&http.Client{})
+	// Region
+	{
+		region, err := metadataClient.Region()
+		if err != nil {
+			return fmt.Errorf("error reading region from Aliyun: %v", err)
+		}
+		a.region = region
+		if a.region == "" {
+			return fmt.Errorf("region metadata was empty")
+		}
+		glog.Infof("Found region=%q", a.region)
+	}
+
+	// Zone
+	{
+		zone, err := metadataClient.Zone()
+		if err != nil {
+			return fmt.Errorf("error reading zone from Aliyun: %v", err)
+		}
+		a.zone = zone
+		if a.zone == "" {
+			return fmt.Errorf("zone metadata was empty")
+		}
+		glog.Infof("Found zone=%q", a.zone)
+	}
+
+	// Instance Name
+	{
+		instanceId, err := metadataClient.InstanceID()
+		if err != nil {
+			return fmt.Errorf("error reading instance ID from Aliyun: %v", err)
+		}
+		a.instanceId = instanceId
+		if a.instanceId == "" {
+			return fmt.Errorf("instance ID metadata was empty")
+		}
+		glog.Infof("Found instanceId=%q", a.instanceId)
+	}
+
+	// Internal IP
+	{
+		internalIP, err := metadataClient.PrivateIPv4()
+		if err != nil {
+			return fmt.Errorf("error querying InternalIP from Aliyun: %v", err)
+		}
+		if internalIP == "" {
+			return fmt.Errorf("InternalIP from metadata was empty")
+		}
+		a.internalIP = net.ParseIP(internalIP)
+		if a.internalIP == nil {
+			return fmt.Errorf("InternalIP from metadata was not parseable(%q)", internalIP)
+		}
+		glog.Infof("Found internalIP=%q", a.internalIP)
+	}
+
+	// Cluster Tag
+	{
+		describeTagsArgs := &ecs.DescribeTagsArgs{
+			RegionId:     common.Region(a.region),
+			ResourceType: ecs.TagResourceInstance,
+			ResourceId:   a.instanceId,
+		}
+		result, _, err := a.client.DescribeTags(describeTagsArgs)
+		if err != nil {
+			return fmt.Errorf("error querying Aliyun instance tags: %v", err)
+		}
+		for _, tag := range result {
+			if tag.TagKey == aliup.TagClusterName {
+				a.clusterTag = tag.TagValue
+			}
+		}
+		if a.clusterTag == "" {
+			return fmt.Errorf("cluster tag metadata was empty")
+		}
+	}
+
+	return nil
+}
+
+// AttachVolume attaches the specified volume to this instance, returning the mountpoint & nil if successful
+func (a *ALIVolumes) AttachVolume(volume *Volume) error {
+	// TODO: what if this volume has already been attached to another instance?
```

Comment:
That's actually good - this acts as a sort of mutex for leader election.  We do assume that the cloud API won't accidentally let us make two attachments, even if we make concurrent "attach volume" calls.